feat(login): Implement refresh tokens

JustinNgxyen opened the pull request:
**Synopsis**:
Created refresh tokens that allow applicants to gain new access tokens without making the user re-login.

**Milestones**:
- The login route now creates both a refresh token and an access token when a user logins. The refresh token has a longer expiration date than the access token so that when the access token expires, the refresh token can create a new access token. 
- Logic of generating and verifying tokens moved to token.js
- The refresh and access tokens each have their own secret values from the environment variables.
- Each token will be verified according to their own separate secrets established in .env.
- The token route will make sure the refresh token is in the token list, as well as if the refresh token has been revoked, before updating the list with a new access token if the old one has expired.
- The revokeToken route will add the refresh token to a blacklist using Redis.

**Testing**:
- Postman was used to see if access and refresh tokens are created when a POST request to the login route is made using an email and password.
- The refresh token is used along with the email to create a POST request to the token route to test if an access token is created. 
- A GET request is made to the user route to test whether an access token contains the correct properties of the payload (user id, first name, last name, etc.). It also will see if the access token has expired.

RFC: https://docs.google.com/document/d/12DvNiTidBfp54obyKWnv8F2Bgs79_ZJzODNnsffAyzk/edit?usp=sharing

JustinNgxyen committed:
began on the refresh token route

JustinNgxyen committed:
updated newToken and refreshToken using jwt.sign. Fixed token route so that a new token is created when old one has expired during logged-in session

JustinNgxyen committed:
feat(server): updated code to store refresh tokens

JustinNgxyen committed:
Merge branch 'Justin/login_auth2' of https://github.com/cppsea/icebreak into Justin/login_auth2

JustinNgxyen committed:
updated code to store refresh tokens

JustinNgxyen committed:
created refresh and access tokens

JustinNgxyen committed:
feat (auth): updated client secret name

JustinNgxyen committed:
feat(auth): updated routes to use token.js

JustinNgxyen committed:
feat(auth): Implemented refresh token rotation

JustinNgxyen committed:
feat(auth): Implemented redis for revoking tokens

JustinNgxyen committed:
feat(auth): Add revoked tokens to blacklist

JustinNgxyen committed:
logic of generating/verifying tokens to token.js

JustinNgxyen committed:
token route checks if refresh token is revoked

lxkedinh commented on package.json:
```diff
@@ -30,5 +30,9 @@
+  "dependencies": {
+    "ioredis": "^5.3.2",
+    "redis": "^4.6.7"
+  }
```
These dependencies shouldn't be here. Remember that this project is a monorepo that contains both the frontend and backend repos. Remove these and then install these dependencies in `packages/server`.

lxkedinh commented on packages/server/routes/api/auth.js:
```diff
@@ -50,19 +54,25 @@ router.post("/verify", async (request, response) => {
+    if (request.user.exp < currentTimestamp) {
+      throw new Error("Access token has expired");
+    }
+    
```
This code to check expired tokens is unnecessary since the JWT package we're using already has expiration checking built-in. Try looking into incorporating that into the error handling for the `authenticate()` function in auth controllers.

https://github.com/auth0/node-jsonwebtoken#tokenexpirederror

lxkedinh commented on packages/server/routes/api/auth.js:
```diff
@@ -145,36 +158,56 @@ router.post("/register", async (request, response) => {
+        });
+
+        await postgres.connect(); 
+
```
Why are you initializing postgres client here? The postgres client is initialized in a separate utility file and then imported everywhere we need it, so we don't reinitialize it every time we access the database.

lxkedinh commented on packages/server/routes/api/auth.js:
```diff
@@ -5,12 +5,16 @@ const token = require("../../utils/token");
 const { OAuth2Client } = require("google-auth-library");
 const bcrypt = require("bcrypt");
 const uniqid = require("uniqid");
+const tokenList = {};
```
`tokenList` shouldn't be needed anymore in this file since we're using Redis to keep track of blacklisted tokens.

lxkedinh commented on packages/server/routes/api/auth.js:
```diff
@@ -196,24 +229,132 @@ router.post("/login", async (request, response) => {
+  }
+});
+
+router.post('/revokeToken', async (request, response) => {
```
This should be `/revoke-token` to follow REST API naming convention. We should follow this when naming new API routes from now on, I'll add this to the resources channel.

![image](https://github.com/cppsea/icebreak/assets/36950420/5fda7807-4c59-4dbc-8fd4-e03d010399ad)

https://restfulapi.net/resource-naming/

lxkedinh commented on packages/server/routes/api/auth.js:
```diff
@@ -196,24 +229,132 @@ router.post("/login", async (request, response) => {
+      token.verifyRefreshToken(refreshToken);
+      
+          // Check if the refresh token exists in tokenList
+          if (refreshToken in tokenList) {
```
You should change this logic since we will be using Redis to reference against a blacklist of revoked tokens instead of checking if tokens are in `tokenList`.

lxkedinh commented on packages/server/utils/token.js:
```diff
@@ -1,16 +1,40 @@
+    last_login: user.last_login,
+    first_name: user.first_name,
+    last_name: user.last_name,
+    avatar: user.avatar,
```
We should try to keep the size of our tokens as small as possible. We should discuss with the frontend team to see which user properties are necessary to have in tokens and which ones are ok to leave out.

JustinNgxyen committed:
updated changes to pr

lxkedinh commented on packages/server/controllers/auth.js:
```diff
@@ -102,9 +104,35 @@ async function deserialize(id, callback) {
-  next();
+  
+  try{
+  token.verifyAccessToken(authToken);
+
+  jwt.verify(authToken, process.env.WEB_CLIENT_SECRET, function(err, decoded) {
+    if (err) {
+      if (err.name === 'TokenExpiredError') {
+        // Token has expired
+        response.status(401).json({ 
+          error: 'Access token has expired' 
+        });
+      } else {
+        // Other token verification errors
+        response.status(401).json({ 
+          error: 'Invalid access token' 
+        });
+      }
+    } else {
+      // Token is valid
+      request.user = decoded;
+      next();
+    }
+  });
```
Why do you have `token.verifyAccessToken(authToken);` and then verify the token again in the next line? You can move this logic to `verifyAccessToken` and throw errors where the `response.status().json()` methods are. That way, the errors you throw in `verifyAccessToken` will be caught and handled in this try/catch block instead.

JustinNgxyen committed:
updated changes to token.js

JustinNgxyen committed:
fixed expiration date of access tokens

JustinNgxyen committed:
/user will return the payload of the access token

JustinNgxyen committed:
updated verifyAccessToken

lxkedinh committed:
chore: merge new changes from main

lxkedinh committed:
Merge branch 'main' into Justin/login_auth2

lxkedinh committed:
chore: merged with main

lxkedinh committed:
chore: refactored callbacks to promises
- changed some variable names to be clearer
- added guard clauses to some auth api routes

lxkedinh committed:
feat: cleaned up authorization flow
- moved some logic from route handlers to auth controller
- adjusted db schema

lxkedinh committed:
Merge branch 'main' into Justin/login_auth2
